### PR TITLE
resource info spec compliance, instrumentation attribute

### PR DESCRIPTION
**Pull request by brettmc**
* When merging resources, later (updating) keys should override existing keys, per spec. Update to work this way, and adjust some tests accordingly.
* Reorganise detectors so that environment runs last (per spec, OTEL_SERVICE_NAME should be the highest priority way of setting service name)
* Adding the experimental telemetry.auto.version attribute, if the auto-instrumentation extension is installed.

### CODE REVIEW

File: src/SDK/Resource/ResourceInfoFactory.php [hunk 1]
```diff
@@ -13,17 +13,17 @@
...
-     * @return ResourceInfo
+     * @see https://github.com/open-telemetry/opentelemetry-specification/blob/v1.20.0/specification/resource/sdk.md#merge
      */
     public static function merge(ResourceInfo ...$resources): ResourceInfo
     {
         $attributes = [];
 
         foreach ($resources as $resource) {
-            $attributes += $resource->getAttributes()->toArray();
+            $attributes = array_merge($attributes, $resource->getAttributes()->toArray());
```
Comment (Nevay) [inline src/SDK/Resource/ResourceInfoFactory.php:26]: Can't use `array_merge`, the attributes array can contain numeric keys.

**Review by Nevay — COMMENTED** [review]
IMO we should just specify that the resources are merged with descending priority, the spec doesn't dictate the argument order.

File: src/SDK/Resource/ResourceInfoFactory.php (hunk 1 above)
Comment (brettmc) [reply src/SDK/Resource/ResourceInfoFactory.php:26]: > the spec doesn't dictate the argument order

I think it does..."old" and "updating" resource params are listed in that order:

> Required parameters:
> * the old resource
> * the updating resource whose attributes take precedence

https://github.com/open-telemetry/opentelemetry-specification/blob/v1.20.0/specification/resource/sdk.md#merge

And in the case that more than two Resources are provided, I think it then follows that later ones take precedence.

> attributes array can contain numeric keys

https://github.com/open-telemetry/opentelemetry-specification/blob/v1.20.0/specification/common/README.md#attribute says `The attribute key MUST be a non-null and non-empty string` - perhaps we need to cast numeric keys to strings, which should resolve not being able to use `array_merge`... It's certainly not covered by any tests, and probably should be. What do you think?

Comment (brettmc) [reply src/SDK/Resource/ResourceInfoFactory.php]: @Nevay you were right about `array_merge` (I didn't realise PHP forces numeric-looking keys to numbers internally), so I've instead applied array_reverse to deal with precedence. I added some tests around numeric keys + merging.

File: tests/Unit/SDK/Common/Attribute/AttributesTest.php [hunk 2]
```diff
@@ -41,6 +41,16 @@ public function test_numeric_attribute_name(): void
         }
     }
 
+    public function test_integer_attribute_key_is_cast_to_string(): void
```
Comment (Nevay) [inline tests/Unit/SDK/Common/Attribute/AttributesTest.php:44]: Test above (`test_numeric_attribute_name`) can be removed, duplicates this test.

**Review by Nevay — APPROVED** [review]

File: src/SDK/Resource/ResourceInfoFactory.php (hunk 1 above)
Comment (Nevay) [reply src/SDK/Resource/ResourceInfoFactory.php]: Does the spec specify that the parameter order must/should be followed? I always assumed that this is entirely up to the implementation.
(fwiw I'm using `Resource::merge(...$this->resources, _default: Resource::getDefault())` in my implementation which can't be written as concisely with ascending priority.)

Comment (brettmc) [reply src/SDK/Resource/ResourceInfoFactory.php]: Not exactly, but it does have this `MUST`:

> `If a key exists on both the old and updating resource, the value of the updating resource MUST be picked (even if the updated value is empty)`

To me, that's saying that if you start with a resource, then merge a second resource, then the second resource's keys take precedence and should overwrite the first.

Looking at other language implementations, `merge` is usually a method on a resource eg `$resourceOne->merge($resourceTwo)` which I think is clearer as to which is the old and the updating resource, and the spec was probably written with that in mind.

**Comment by brettmc** [comment]
Sorry, I broke my forked repo and started over, and this got closed when I deleted my fork. Recreated in #989

**Pull request closed without merging**